### ISSUE TITLE
Mandated minimum for CL_DEVICE_DOUBLE_FP_CONFIG

### ISSUE THREAD

**Issue by kpet** [KhronosGroup/OpenCL-Docs#673]
The 1.0, 1.1 and 1.2 specifications state the minimum set of capabilities is:
```
CL_FP_FMA | 
CL_FP_ROUND_TO_NEAREST | 
CL_FP_ROUND_TO_ZERO | 
CL_FP_ROUND_TO_INF | 
CL_FP_INF_NAN | 
CL_FP_DENORM
```

The 2.0, 2.1, 2.2 and latest (at issue creation time) unified specifications state:

```
CL_FP_FMA | 
CL_FP_ROUND_TO_NEAREST | 
CL_FP_INF_NAN | 
CL_FP_DENORM
```

The CTS requires:

```
CL_FP_FMA |
CL_FP_ROUND_TO_NEAREST |
CL_FP_ROUND_TO_ZERO |
CL_FP_ROUND_TO_INF |
CL_FP_INF_NAN |
CL_FP_DENORM
```
regardless of the version.

Was this change intentional?
- If yes, we should update the unified specification to document that the minimum set of required capabilities has changed in 2.0 as well as the CTS test to require the right capabilities for each version
- If no, we should fix the 2.0 and onwards specifications.

**Comment by alycm** [comment]
I didn't find anything on legacy bugzilla suggesting this change was deliberate, issue 6363 touches on the related issue that removing (deprecating?) `__ROUNDING_MODE__` in OpenCL 1.1 made `CL_FP_ROUND_TO_INF` unreachable, and that `CL_FP_ROUND_TO_ZERO` is only reachable in embedded profile.

But that doesn't even fit the 2.0 switchover point here.

**Comment by bashbaug** [comment]
I didn't find anything in my spec archaeology either, though I do see that this changed between rev5 and rev6 of the OpenCL 2.0 draft specs, so sometime around April 2013.

Note, the "minimum requirements" for these floating-point capabilities are a little strange because very few are functional:

* `CL_FP_DENORM` is, otherwise denormals are not supported.
* For the embedded profile, whether `CL_FP_ROUND_TO_NEAREST` is supported defines the default rounding mode.
* Also for the embedded profile, `CL_FP_INF_NAN` defines whether infinities and NaNs are supported.

For the others though, and for the full profile, I think they're all purely informational?  For example, all devices need to support an `fma()` built-in function regardless of `CL_FP_FMA`?

**Comment by Jeremy-Kemp** [comment]
> For the others though, and for the full profile, I think they're all purely informational?

I'm also wondering what an application could actually do with this information, specifically in the case where an implementation exposed multiple rounding modes. Aside from the explicit conversion built-ins, it isn't possible to select between different rounding modes since the deprecation of `cl_khr_select_fprounding_mode`.

**Comment by bashbaug** [comment]
Thanks for the feedback @Oblomov.  There has been an internal Khronos issue open about this for as long as I can remember, and if I'm not mistaken it has a [link](http://wok.oblomov.eu/tecnologia/gpgpu/opencl-rounding-modes/) to one of your blog posts about this topic from 2014 (!).

IIRC we usually get hung up on:

* Which functions ("instructions") should be supported?
* Which data types should be supported for these functions (e.g. half? float? double?)?
* Which rounding modes should be required?  Should any be optional?

My notes usually say that "this would be an execllent vendor or EXT extension!"  I'd certainly encourage reaching out to the hardware vendor(s) with per-instruction rounding mode control to see if they would be interested in defining such an extension.  Then, if it proves to be popular, we could consider standardizing it with much higher confidence.

**Comment by Oblomov** [comment]
Thanks for the reply @bashbaug. And yes, this has been a pet peeve of mine for a looong time ;-)

Of course as a user it would be ideal to have support for all rounding modes for all built-ins and all (device-supported) data types, but I can see why this could be a problem for implementations.

I wonder if this could be made more palatable to implementors by providing a permissively licensed header-only library to compensate for missing rounding modes or instructions? The extension could then make some of them optional, with appropriate device properties (e.g. the math capabilities being discussed here) and defines that could be used by the library to determine which overrides need to be implemented.

**Comment by bashbaug** [comment]
> I wonder if this could be made more palatable to implementors by providing a permissively licensed header-only library to compensate for missing rounding modes or instructions?

Yes absolutely, this would be great!  I think it'd also help to distinguish between operations that could feasibly be provided by a library and those that should be provided directly by an implementation instead.

I am not a numerics expert, so this may be a silly question, but for my own education if nothing else: Is the maximal set of operations that would benefit from additional rounding mode controls the operations that are currently "correctly rounded"?  If so, perhaps the list isn't _too_ long?

https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#_ulp_values_for_math_instructions_full_profile

**Comment by Oblomov** [comment]
I would argue that that's the _minimal_ set that should support rounding mode control (including div and sqrt for fp32, of course, if the corresponding flag is set in the device FP config for single-precision).

The maximal set would be “all of them”, even though still with the given accuracy, so that e.g. you would be sure that the `cos` will return an underestimation of the exact result (by up to 4 ULP) when the rounding mode is towards zero, and an overestimation of it (again, by up to 4 ULP) when the rounding mode is away from zero.

**Comment by jlewis-austin** [comment]
> you would be sure that the `cos` will return an underestimation of the exact result (by up to 4 ULP) when the rounding mode is towards zero, and an overestimation of it (again, by up to 4 ULP) when the rounding mode is away from zero.

This would be a new invention, as rounding modes are defined to only apply to the last bit of correctly-rounded functions. What you're describing is more at the algorithmic level and would require redesign to tighten the accuracy specs in order to accommodate introducing artificial bias into the results. That could be useful for a library, but it would be an inaccurate usage of the term "rounding mode".

**Comment by Oblomov** [comment]
While I agree it's a stretch of the formal definition, I'm not sure why you talk about _artificial_ bias. The accuracy constraints effectively tell you how many bits of the numerical result should match the bits of the exact result. In the worst case, with directed rounding you can work as if you only had that many bits in the mantissa, and then pad the rest with 0 or 1 depending on sign and rounding direction. It isn't any better of worse than the meaninglessness of the last bits in the normal case.

**Comment by jlewis-austin** [comment]
Maybe _artificial_ is the wrong word; I just mean bias that is significantly higher than what is introduced from these rounding modes which are defined in terms of last-bit rounding. My main objection is the stretching of the formal definition of the rounding modes, since they're precisely defined, very entrenched and function very differently from what you're describing. There are correctly-rounded libraries in existence that would have a stronger claim to use of the rounding terminology if somebody ever decided to use them in an OpenCL implementation.
>  It isn't any better of worse than the meaninglessness of the last bits in the normal case.

The last bits are actually critical in meeting the ulp spec. We can't assume "good" and "bad" bits, because ALL bits can differ from the exact result and still meet the spec. A <1 ulp spec doesn't mean the lowest bit is a don't care--it means there are exactly two acceptable output values for any input (and just one if the true result is exactly representable). For example, if the true value lies between `0x7fff and 0x8000`, those are the only acceptable results (I'm assuming an integer output for simplicity). `0x7ffe or 0x8001` would both be out of spec, even though it's just the last bit that's changed. 

This gets more complicated when you get to specs like <5  ulp. That translates to a fractional number of bits, so even if it worked to set LSBs to 1 or 0, how many bits would you set in that case?

The only reliable way to use an existing algorithm to behave like you describe would be to add/subtract 4 ulp to every output. That would effectively shift the error spec from (-4, 4) ulp to (0, 8) ulp/(-8, 0) ulp. This guarantees that all results will be either correct or overestimated, while not affecting the inherent accuracy of the algorithm. Unfortunately it no longer meets the original spec requirement.

**Comment by Oblomov** [comment]
I see what you mean now, but I don't think that's really the case (and yes, I did express myself poorly about the bits of precision, due to the effect of carry).  Directed rounding doesn't introduce more bias. Of course it does if you implement it naively by adding/subtracting the ULP, but that's not what the spec should mandate: the error constraints should remain the same, but only be “laterla”, just like what happens for correctly rounded functions.

For correctly rounded functions, the error spec is (-1, 1) ulp for RNE, but with directed rounding it becomes either (-1,0] or [0, 1) (depending on rounding mode). The same would be true for higher ULPs: under directed rounding, the spec would require the result to be within (n, 0] or [0, n) of the correct value, when with RNE it's in (-n, n).

And yes, this _may_ require the algorithm implementing the directed rounding version be different from the one used for RNE, but not necessarily with stricter error constraints.
